// @elgato-stream-deck/core/dist/id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODEL_NAMES = exports.DeviceModelId = void 0;
var DeviceModelId;
(function (DeviceModelId) {
    DeviceModelId["ORIGINAL"] = "original";
    DeviceModelId["ORIGINALV2"] = "originalv2";
    DeviceModelId["ORIGINALMK2"] = "original-mk2";
    DeviceModelId["MINI"] = "mini";
    DeviceModelId["XL"] = "xl";
    DeviceModelId["PEDAL"] = "pedal";
    DeviceModelId["PLUS"] = "plus";
    DeviceModelId["NEO"] = "neo";
    DeviceModelId["STUDIO"] = "studio";
})(DeviceModelId || (exports.DeviceModelId = DeviceModelId = {}));
exports.MODEL_NAMES = {
    [DeviceModelId.ORIGINAL]: 'Stream Deck',
    [DeviceModelId.MINI]: 'Stream Deck Mini',
    [DeviceModelId.XL]: 'Stream Deck XL',
    [DeviceModelId.ORIGINALV2]: 'Stream Deck',
    [DeviceModelId.ORIGINALMK2]: 'Stream Deck MK.2',
    [DeviceModelId.PLUS]: 'Stream Deck +',
    [DeviceModelId.PEDAL]: 'Stream Deck Pedal',
    [DeviceModelId.NEO]: 'Stream Deck Neo',
    [DeviceModelId.STUDIO]: 'Stream Deck Studio',
};
