// @elgato-stream-deck/core/dist/models/mini.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamDeckMiniFactory = StreamDeckMiniFactory;
const generic_gen1_js_1 = require("./generic-gen1.js");
const id_js_1 = require("../id.js");
const controlsGenerator_js_1 = require("../controlsGenerator.js");
const imageWriter_js_1 = require("../services/imageWriter/imageWriter.js");
const headerGenerator_js_1 = require("../services/imageWriter/headerGenerator.js");
const miniProperties = {
    MODEL: id_js_1.DeviceModelId.MINI,
    PRODUCT_NAME: id_js_1.MODEL_NAMES[id_js_1.DeviceModelId.MINI],
    SUPPORTS_RGB_KEY_FILL: false, // TODO - verify this
    CONTROLS: (0, controlsGenerator_js_1.freezeDefinitions)((0, controlsGenerator_js_1.generateButtonsGrid)(3, 2, { width: 80, height: 80 })),
    KEY_SPACING_HORIZONTAL: 28,
    KEY_SPACING_VERTICAL: 28,
    FULLSCREEN_PANELS: 0,
};
function StreamDeckMiniFactory(device, options) {
    return (0, generic_gen1_js_1.StreamDeckGen1Factory)(device, options, miniProperties, new imageWriter_js_1.StreamdeckDefaultImageWriter(new headerGenerator_js_1.StreamdeckGen1ImageHeaderGenerator()), { colorMode: 'bgr', rotate: true, yFlip: true }, 2835);
}
