// @elgato-stream-deck/core/dist/models/originalv2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamDeckOriginalV2Factory = StreamDeckOriginalV2Factory;
const base_js_1 = require("./base.js");
const generic_gen2_js_1 = require("./generic-gen2.js");
const id_js_1 = require("../id.js");
const controlsGenerator_js_1 = require("../controlsGenerator.js");
const origV2Properties = {
    MODEL: id_js_1.DeviceModelId.ORIGINALV2,
    PRODUCT_NAME: id_js_1.MODEL_NAMES[id_js_1.DeviceModelId.ORIGINALV2],
    SUPPORTS_RGB_KEY_FILL: false, // TODO - verify SUPPORTS_RGB_KEY_FILL
    CONTROLS: (0, controlsGenerator_js_1.freezeDefinitions)((0, controlsGenerator_js_1.generateButtonsGrid)(5, 3, { width: 72, height: 72 })),
    KEY_SPACING_HORIZONTAL: 25,
    KEY_SPACING_VERTICAL: 25,
    FULLSCREEN_PANELS: 0,
    HAS_NFC_READER: false,
    SUPPORTS_CHILD_DEVICES: false,
};
function StreamDeckOriginalV2Factory(device, options) {
    const services = (0, generic_gen2_js_1.createBaseGen2Properties)(device, options, origV2Properties, null);
    return new base_js_1.StreamDeckBase(device, options, services);
}
