// @elgato-stream-deck/core/dist/models/xl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamDeckXLFactory = StreamDeckXLFactory;
const base_js_1 = require("./base.js");
const generic_gen2_js_1 = require("./generic-gen2.js");
const id_js_1 = require("../id.js");
const controlsGenerator_js_1 = require("../controlsGenerator.js");
const xlProperties = {
    MODEL: id_js_1.DeviceModelId.XL,
    PRODUCT_NAME: id_js_1.MODEL_NAMES[id_js_1.DeviceModelId.XL],
    SUPPORTS_RGB_KEY_FILL: false, // rev2 doesn't support it, even though rev1 does
    CONTROLS: (0, controlsGenerator_js_1.freezeDefinitions)((0, controlsGenerator_js_1.generateButtonsGrid)(8, 4, { width: 96, height: 96 })),
    KEY_SPACING_HORIZONTAL: 32,
    KEY_SPACING_VERTICAL: 39,
    FULLSCREEN_PANELS: 0,
    HAS_NFC_READER: false,
    SUPPORTS_CHILD_DEVICES: false,
};
function StreamDeckXLFactory(device, options) {
    const services = (0, generic_gen2_js_1.createBaseGen2Properties)(device, options, xlProperties, null);
    return new base_js_1.StreamDeckBase(device, options, services);
}
